### PR TITLE
Timed loop

### PR DESCRIPTION
**Pull request by mreiger**

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -3,160 +3,257 @@ package main
...
+	cmd.Flags().StringP("kubecfg", "k", homedir+"/.kube/config", "kubecfg path to the cluster to account")
+
+	cmd.Flags().StringP("namespace", "", "kube-system", "the namespace of the audit-tailer service")
+	cmd.Flags().StringP("service-name", "", "kubernetes-audit-tailer", "the service name of the audit-tailer service")
+	cmd.Flags().IntP("service-port", "", 24224, "the service port of the audit-tailer service")
+	cmd.Flags().StringP("check-schedule", "", "*/1 * * * *", "cron schedule when to check for service changes")
+
+	err = viper.BindPFlags(cmd.Flags())
+	if err != nil {
+		logger.Fatalw("unable to construct root command", "error", err)
```
Comment (Gerrit91) [inline main.go]: The `logger` will still be nil at this point in time, I think.

File: main.go [hunk 2]
```diff
@@ -3,160 +3,257 @@ package main
...
+func run(opts *Opts) error {
+	logger.Infow("Options", "opts", opts)
+	// initialise our synchronisation channels
+	forwarderKilledChan = make(chan struct{})
+	killForwarderChan = make(chan struct{}, 1)
+
+	// Prepare K8s
+	client, err := loadClient(opts.KubeCfg)
+	if err != nil {
+		logger.Fatalw("Unable to connect to k8s", "Error", err)
```
Comment (Gerrit91) [inline main.go]: After the entire init phase there should be no need to call `Fatal` anymore, just return the error and let the underlying layer handle the issue.

File: main.go [hunk 3]
```diff
@@ -166,3 +263,86 @@ func loadClient(kubeconfigPath string) (*k8s.Clientset, error) {
...
+		if targetService != nil { // This means a service was previously seen, and a forwarder should already be running.
+			logger.Infow("Service went away, killing forwarder")
+			killForwarder()
+			targetService = nil
+		}
+		return err
+	}
+	logger.Infow("Service gotten", "service", service)
+	serviceIP := service.Spec.ClusterIP
+	servicePort := service.Spec.Ports[0].Port
```
Comment (Gerrit91) [inline main.go:287]: Possible index out of bounds

Comment (mreiger) [reply main.go:287]: Looks like there must be at least one servicePort in a kubernetes service; trying to deploy one with empty spec: ports: [] gives:

```
kubectl apply -f kubernetes-audit-tailer-mod2.yaml
deployment.apps/kubernetes-audit-tailer configured
configmap/kubernetes-audit-tailer-config unchanged
The Service "kubernetes-audit-tailer" is invalid: spec.ports: Required value
```
Checking the length of the Ports slice(?) anyway, just to be sure.

File: main.go (hunk 2 above)
Comment (mreiger) [reply main.go]: Changed - not sure if this is what you had in mind?

File: main.go [hunk 4]
```diff
@@ -3,160 +3,257 @@ package main
...
+	}
+
+	cmd.Flags().StringP("log-level", "", "info", "sets the application log level")
+	cmd.Flags().StringVarP(&cfgFile, "config", "c", "", "alternative path to config file")
+
+	cmd.Flags().StringP("kubecfg", "k", homedir+"/.kube/config", "kubecfg path to the cluster to account")
+
+	cmd.Flags().StringP("namespace", "", "kube-system", "the namespace of the audit-tailer service")
+	cmd.Flags().StringP("service-name", "", "kubernetes-audit-tailer", "the service name of the audit-tailer service")
+	cmd.Flags().IntP("service-port", "", 24224, "the service port of the audit-tailer service")
```
Comment (mwindower) [inline main.go:109]: service-port flag seems to be unused

File: main.go (hunk 2 above)
Comment (mreiger) [reply main.go]: Now doing return(err)

**Comment by mreiger** [comment]
Merging this now and making a new branch for TLS integration.

File: main.go (hunk 4 above)
Comment (mreiger) [reply main.go:109]: Thanks - removed!

**Pull request merged**